### PR TITLE
[sdk/dotnet] Check for `pulumiplugin.json` in package lookup

### LINKED ISSUE #7244: .NET: Support custom server URLs for plugins
.NET doesn't currently have a way to indicate that a package has a plugin from a custom server URL.

- [codegen] Adopt `pulumiplugin.json` to support custom server URLs (this is the same file used by Python packages in the same format as the `pulumi` element in `package.json` for Node.js).
- [langhost] Allow non `Pulumi.` prefixed packages when searching for packages that need plugins in `GetRequiredPlugins`.

### PR DESCRIPTION
**Pull request by iwahbe**
# Description

Fixes #7244

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go [hunk 1]
```diff
@@ -298,55 +299,104 @@ func (host *dotnetLanguageHost) DetermineDotnetPackageDirectory(
...
+		version: version,
+	}
+}
+
+func DeterminePluginDependency(packageDir, packageName, packageVersion string) (*pulumirpc.PluginDependency, error) {
 
 	logging.V(5).Infof("GetRequiredPlugins: Determining plugin dependency: %v, %v, %v",
 		packageDir, packageName, packageVersion)
 
 	// Check for a `~/.nuget/packages/package_name/package_version/content/version.txt` file.
```
Comment (justinvp) [inline sdk/dotnet/cmd/pulumi-language-dotnet/main.go:333]: Nit: Maybe update this comment to mention `pulumiplugin.json` as well?

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go [hunk 2]
```diff
@@ -298,55 +299,104 @@ func (host *dotnetLanguageHost) DetermineDotnetPackageDirectory(
...
-			// just ignore it.
-			logging.V(5).Infof("GetRequiredPlugins: No such file")
-			return nil, nil
-		}
-
-		// some other error.  report it as it means we can't read this important file.
-		logging.V(5).Infof("GetRequiredPlugins: err: %v", err)
+	pulumiPlugin, err := plugin.LoadPulumiPluginJSON(pulumiPluginFilePath)
+	if os.IsNotExist(err) {
+		pulumiPlugin = nil
```
Comment (justinvp) [inline sdk/dotnet/cmd/pulumi-language-dotnet/main.go:343]: Is this necessary? Won't `pulumiPlugin` already be `nil` when `err != nil` and therefore `os.IsNotExists(err)` is `true`?

**Review by justinvp — APPROVED** [review]
LGTM

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go (hunk 2 above)
Comment (iwahbe) [reply sdk/dotnet/cmd/pulumi-language-dotnet/main.go]: We need the check, but not the assignment. I can fold it down to this
```go
if err != nil && !os.IsNotExist(err) {
		return nil, err
	}
```

**Pull request merged**